### ISSUE TITLE
Signing failed with compatible "IAS/ECC Morpho MinInt - Agent Card"

### ISSUE THREAD

**Issue by redsectors** [OpenSC/OpenSC#1096]
### Expected behaviour

Hello,

Having this card:
http://www.imprimerienationale.fr/en/products-services/agent-and-professional-cards.html

I should be able to sign data using pkc11-tool and SHA1-RSA-PKCS or SHA256-RSA-PKCS algorithm.

Dragonous:OpenSC root# opensc-tool -a
Using reader with a card: Alcor Micro AU9560
3b:df:96:00:80:31:fe:45:00:31:b8:64:04:1f:ec:c1:73:94:01:80:82:90:00:ec
Dragonous:OpenSC root# pkcs11-tool -M
Using slot 0 with a present token (0x0)
Supported mechanisms:
  SHA-1, digest
  SHA256, digest
  SHA384, digest
  SHA512, digest
  MD5, digest
  RIPEMD160, digest
  GOSTR3411, digest
  RSA-PKCS, keySize={1024,2048}, hw, decrypt, sign, verify
  SHA1-RSA-PKCS, keySize={1024,2048}, sign, verify
  SHA256-RSA-PKCS, keySize={1024,2048}, sign, verify
  RSA-PKCS-KEY-PAIR-GEN, keySize={1024,2048}, generate_key_pair

### Actual behaviour

Only RSA-PKCS signing works. Encryption works fine also.
SHA1-RSA-PKCS or SHA256-RSA-PKCS (used by outlook ?) do not work.
I got CKR_MECHANISM_INVALID error.

### Steps to reproduce

Dragonous:OpenSC root# pkcs11-tool -s -m RSA-PKCS
Using slot 0 with a present token (0x0)
Logging in to "PIN_Global (pr_XXXXXXX".
Please enter User PIN: 
Using signature algorithm RSA-PKCS
blabla

--> OK

Dragonous:OpenSC root# pkcs11-tool -s -m SHA1-RSA-PKCS
Using slot 0 with a present token (0x0)
Logging in to "PIN_Global (pr_XXXXXXX".
Please enter User PIN: 
Using signature algorithm SHA1-RSA-PKCS
blabla
error: PKCS11 function C_SignInit failed: rv = CKR_MECHANISM_INVALID (0x70)
Aborting.
Dragonous:OpenSC root# 

### Logs

```
0x7fffc8e8e3c0 15:30:28.791 [opensc-pkcs11] pkcs11-object.c:410:C_FindObjectsInit: 4 matching objects
0x7fffc8e8e3c0 15:30:28.791 [opensc-pkcs11] misc.c:280:session_get_operation: called
0x7fffc8e8e3c0 15:30:28.791 [opensc-pkcs11] misc.c:280:session_get_operation: called
0x7fffc8e8e3c0 15:30:28.791 [opensc-pkcs11] pkcs11-object.c:59:sc_find_release: freeing 32 handles used 4  at 0x7fbe3a8045f0
0x7fffc8e8e3c0 15:30:31.331 [opensc-pkcs11] framework-pkcs15.c:3465:pkcs15_prkey_get_attribute: pkcs15_prkey_get_attribute() called
0x7fffc8e8e3c0 15:30:31.331 [opensc-pkcs11] framework-pkcs15.c:3465:pkcs15_prkey_get_attribute: pkcs15_prkey_get_attribute() called
0x7fffc8e8e3c0 15:30:31.331 [opensc-pkcs11] mechanism.c:250:sc_pkcs11_sign_init: called
0x7fffc8e8e3c0 15:30:31.331 [opensc-pkcs11] mechanism.c:256:sc_pkcs11_sign_init: mechanism 0x6, key-type 0x0
0x7fffc8e8e3c0 15:30:31.331 [opensc-pkcs11] misc.c:258:session_start_operation: called
0x7fffc8e8e3c0 15:30:31.331 [opensc-pkcs11] misc.c:259:session_start_operation: Session 0x7fbe3941d280, type 1
0x7fffc8e8e3c0 15:30:31.331 [opensc-pkcs11] mechanism.c:367:sc_pkcs11_signature_init: called
0x7fffc8e8e3c0 15:30:31.331 [opensc-pkcs11] mechanism.c:386:sc_pkcs11_signature_init: returning with: 112
0x7fffc8e8e3c0 15:30:31.331 [opensc-pkcs11] mechanism.c:274:sc_pkcs11_sign_init: returning with: 112
0x7fffc8e8e3c0 15:30:31.331 [opensc-pkcs11] pkcs11-object.c:653:C_SignInit: C_SignInit() = CKR_MECHANISM_INVALID
```

Any idea what can be done?

BR

**Comment by af-anssi** [comment]
There exists a previous issue on this topic including https://github.com/OpenSC/OpenSC/issues/334#issuecomment-68059825 that depicts the conditions you have to compel with.
From the logs, it seems you use a key controlled by the "Global PIN", not the "Sign PIN".
Could you try to sign with the key controlled by the "Sign PIN" ?

**Comment by redsectors** [comment]
Thanks for the answer.
As far as i know i only have one PIN to access all certs & keys...
The same pin code is used on windows system to sign/encrypt data with outlook or log through active directory.

I try your suggestion to sign with the sign key and get same error... even with specifying AID.

pkcs11-tool -s -m SHA1-RSA-PKCS  -d <ID of Private RSA Key [Sign]>
pkcs11-tool -s -m RSA-PKCS  --application-id E828BD080FA00000050450524F200104 -d <ID of Private RSA Key [Sign]>

What is strange is that pkcs11-tool -t -l does not try to test the missing algorithms whatever the cert/key id specified...

Dragonous:OpenSC root# pkcs11-tool -t -l
Using slot 0 with a present token (0x0)
Logging in to "PIN_Global (pr_degu7392_150-303-".
Please enter User PIN: 
C_SeedRandom() and C_GenerateRandom():
  seeding (C_SeedRandom) not supported
  seems to be OK
Digests:
  all 4 digest functions seem to work
  MD5: OK
  SHA-1: OK
  RIPEMD160: OK
Signatures (currently only for RSA)
  testing key 0 (Auth) 
  all 4 signature functions seem to work
  testing signature mechanisms:
    RSA-PKCS: OK
  testing key 1 (2048 bits, label=Sign) with 1 signature mechanism
    RSA-PKCS: OK
  testing key 2 (2048 bits, label=Cipher) with 1 signature mechanism
    RSA-PKCS: OK
  testing key 3 (2048 bits, label=OldCipher1) with 1 signature mechanism
    RSA-PKCS: OK
Verify (currently only for RSA)
  testing key 0 (Auth)
    RSA-PKCS: OK
  testing key 1 (Sign) with 1 mechanism
    RSA-PKCS: OK
  testing key 2 (Cipher) with 1 mechanism
    RSA-PKCS: OK
  testing key 3 (OldCipher1) with 1 mechanism
    RSA-PKCS: OK
Unwrap: not implemented
Decryption (currently only for RSA)
  testing key 0 (Auth) 
    RSA-PKCS: OK
  testing key 1 (Sign) 
    RSA-PKCS: OK
  testing key 2 (Cipher) 
    RSA-PKCS: OK
  testing key 3 (OldCipher1) 
    RSA-PKCS: OK
No errors

**Comment by af-anssi** [comment]
We can investigate with you to confirm/solve this issue if you contact me directly. I think you will find my email address, it is not difficult ;-)

**Comment by frankmorgner** [comment]
@af-anssi @redsectors did you find a way to solve this issue? If so, can you document or fix this problem for other users?

**Comment by redsectors** [comment]
Sorry i just came from vacation :)
We discussed with af-anssi however the card is a bit different despite sharing same ATR.

1/ Card has a different AID

static struct sc_aid INIASECC_AID = {
         { 0x4F, 0x52, 0x41, 0x4E, 0x47, 0x45}, 6
 };

2/ Pin length must be 24 bytes with 0x0 padding if needed 

apdu.datalen = 24;
apdu.lc = 24;

3/ I did not investigate signing issue yet

As i am not C developper, I made some dirty hack to make it work. I can share ADPU sent/received by windows driver if someone can help and write a proper patch.

BR

Red.
